File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_05/ConcatMapTest.java
package com.fResult.reactor.ch5_05;

import java.time.Duration;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

class ConcatMapTest {
  @Test
  void concatMap() {
    val data =
        Flux.just(new Pair(1, 300), new Pair(2, 200), new Pair(3, 100))
            .concatMap(id -> delayReplyFor(id.id(), id.delay()));

    StepVerifier.create(data).expectNext(1, 2, 3).verifyComplete();
  }

  private Flux<Integer> delayReplyFor(int i, long delay) {
    return Flux.just(i).delayElements(Duration.ofMillis(delay));
  }

  static record Pair(int id, long delay) {}
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_05/ConcatMapTest.kt
package com.fResult.reactor.ch5_05

import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import kotlin.test.Test
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.toJavaDuration

class ConcatMapTest {
  @Test
  fun concatMap() {
    val numbers = Flux.just(Pair(1, 300L), Pair(2, 200L), Pair(3, 100L))
      .concatMap { id -> delayReplyFor(id.first, id.second) }

    StepVerifier.create(numbers).expectNext(1, 2, 3).verifyComplete()
  }
}

private fun delayReplyFor(n: Int, delay: Long) =
  Flux.just(n).delayElements(delay.milliseconds.toJavaDuration())
